amethyst_animation/src/systems/sampling.rs

```diff
@@ -0,0 +1,202 @@
+    match control.end {
+        EndControl::Normal => match control.after {
+            RestState::Translation(tr) => transform.translation = tr.into(),
+            RestState::Rotation(r) => transform.rotation = r.into(),
+            RestState::Scale(s) => transform.scale = s.into(),
+        },
+        // looping is handled during duration update
+        _ => {}
+    }
+}
+
+/// Update durations, check if the sampler is finished, start new samplers, and check for aborted
+/// samplers.
+///
+/// ## Parameters
+///
+/// - `control`: sampler control object
+/// - `sampler`: sampler reference from control
+/// - `now`: synchronized `Instant` for the current frame
+///
+/// ## Returns
+///
+/// Will return the new state of the sampling, and optionally a new end control state (for looping)
+fn update_duration_and_check(
+    control: &SamplerControl,
+    sampler: &Sampler,
+    now: &Instant,
+) -> (ControlState, Option<EndControl>) {
+    use resources::ControlState::*;
+    // Update state with new duration
+    // Check duration for end of sampling
+    match control.state {
+        // requested sampling => start interpolating
+        Requested => (Running(now.clone(), Duration::from_secs(0)), None),
+
+        // abort sampling => end interpolating
+        Abort => (Done, None),
+
+        // sampling is running, update duration and check end condition
+        Running(last, duration) => {
+            let current_dur = duration + (*now - last);
+            let last_frame = sampler.input.last().cloned().unwrap_or(0.);
+            let duration_secs = duration_to_secs(current_dur);
+            // duration is past last frame of sampling
+            if last_frame != 0. && duration_secs > last_frame {
+                // Check end conditions
+                match control.end {
+                    // Do loop control
+                    EndControl::Loop(Some(i)) if i <= 1 => (Done, Some(EndControl::Normal)),
+                    EndControl::Loop(None) => (
+                        Running(now.clone(), next_duration(last_frame, duration_secs)),
+                        None,
+                    ),
+                    EndControl::Loop(Some(i)) => (
+                        Running(now.clone(), next_duration(last_frame, duration_secs)),
+                        Some(EndControl::Loop(Some(i - 1))),
+                    ),
+                    // All other end cases will be handled during sampling
+                    _ => (Done, None),
+                }
+            } else {
+                // last frame not reached, keep sampling
+                (Running(now.clone(), current_dur), None)
+            }
+        }
+
+        // Done and paused will be handled during sampling
+        ref state => (state.clone(), None),
+    }
+}
+
+fn next_duration(last_frame: f32, mut dur_s: f32) -> Duration {
+    while last_frame != 0. && dur_s > last_frame {
+        dur_s -= last_frame;
```

Comment:
Division should work faster then loop.
And you probably want to work with milliseconds in `u32` (or nanoseconds in `u64`) instead of seconds in `f32`.